docs: add pros and cons for jaxb micorservice (#1721)
digitalservicebund/a2j-rechtsantragstelle: doc/adr/0018-xml-generation-from-xsd.md

That sparks the question if we could do the type generation from XSD by ourselves while only relying on libraries for the TypeScript-to-XML-parsing.
For the initial generation we could even still use a library for the one-time generation.

- Advantages:

  - No microservice needed:
    - No additional infrastructure
    - No interoperability issues between two services: Functions/methods can be adapted more easily when the data structure changes (e.g. name of a variable or adding/removing variables)
  - We would use the same language as in our application.

- Disadvantages:
  - The code base is already very large and will become even larger in the future (e.g. with TGA - Teilgenerischer Ansatz), and the xml generation requires a lot of additional code:
    - The Typescript classes of the XSD schemas
    - New methods and (zod) schemas
    - More complexity in testing

>     - We will have to maintain at least two different XJustiz versions at the same time for interoperability.

Author: What I imagine might happen is that the other side (in our case the court software systems) might not implement the new XJusitz standard immediately when it comes out, and different courts might be faster/slower in implementing it, the same for the software used by lawyers. So we might need to have different REST API versions at the same time.
(here a link to what I mean https://www.freecodecamp.org/news/how-to-version-a-rest-api/)

>     - We will have to maintain at least two different XJustiz versions at the same time for interoperability.

Reviewer: This is from xjustiz.de:

> Neue XJustiz Versionen werden einmal jährlich am 30. April veröffentlicht. Dabei wird am 30. April eines jeden Jahres die neue XJustiz Version gültig und gleichzeitig die XJustiz Preview für das folgende Jahr veröffentlicht.
Zum Beispiel wird am 30. April 2025 die XJustiz Version 3.5.1 gültig, und die XJustiz Preview 3.6 für das nächste Jahr veröffentlicht.

Meaning there's always just one version really "gültig" - yes, we'd have to implement the new one before that but then only switch once it's "gültig".

You are probably right, there probably will be some softwares not yet supporting the new version, but I do not think we'd know that without asking. And optimizing for that sounds a little out-of-our-way to me?

>     - We will have to maintain at least two different XJustiz versions at the same time for interoperability.

Reviewer: Mh, fascinating. If multiple verisons are "gültig" then there must be a register on who (which Gericht) uses which version, right? I mean, we probably won't send multiple versions or the same xml to the same Gericht?
And if there is no such register (which I have never seen), how would we know what Gericht uses what version? I'd think we wouldn't now and would just assume that they're all using the current "gültige" version. Hence, we'd only send one?

So the assumption to validate here is "There is no register where I can look up what version of xjustiz to use when communicating with whom."?

>     - We will have to maintain at least two different XJustiz versions at the same time for interoperability.

Author: So as far as I know there are different ways to handle this and what you mentioned is one of the ways.
More automatic ways of doing this could be
- via a discovery endpoint (GET / version)
- information during the handshake (not sure exactly where, but I think as part of the header?)
- sending something in the latest version, if 400 Bad Request or 415 Unsupported Media Type is the response  then trying another version or see if the response contains some info like "Unsupported XJustiz version: expected 2.0, received 1.0".

>     - We will have to maintain at least two different XJustiz versions at the same time for interoperability.

Reviewer: Nice, yes, all possible options. But all determined by the "other side", so MJP or EGVP-Enterprise, I think? Meaning, depending on what they implemented, we have to oblidge to it. We don't know what is implemented, so we should find out before implementing a possibly complex version-handling on our own, right?

  - If the XJustiz version changes:
    - it may be difficult to find the changes

>     - manual adjustments required

Author: good idea :)

That sparks the question if we could do the type generation from XSD by ourselves while only relying on libraries for the TypeScript-to-XML-parsing.
For the initial generation we could even still use a library for the one-time generation.

- Advantages:

  - No microservice needed:
    - No additional infrastructure
    - No interoperability issues between two services: Functions/methods can be adapted more easily when the data structure changes (e.g. name of a variable or adding/removing variables)
  - We would use the same language as in our application.

- Disadvantages:

>   - The code base is already very large and will become even larger in the future (e.g. with TGA), and the xml generation requires a lot of additional code:

Reviewer 2: ```suggestion
  - The code base is already very large and will become even larger in the future (e.g. with TGA — Teilgenerischer Ansatz ), and the xml generation requires a lot of additional code:
```
Not everyone knows this

    - The Typescript classes of the XSD schemas
    - New methods and (zod) schemas

>     - Testing

Reviewer 3: Why writing test is disadvantages? Write tests is never bad, bad is write bad code.

Author: You are right that writing tests is not a disadvantage - here it is part of the argument that the code base is already very large and that it will continue growing a lot. Furthermore, the approach also means that we need much more testing than with the jaxb libraray because the jaxb library covers a lot of things without having to write new methods for it as mentioned below:
 - Tests to test the new methods and schemas for the xml generation from the classes (jaxb already provides these methods)
 - Tests to check that the result is a valid XJustiz message (may require a mini microservice)

Reviewer 3: I understand that increasing the codebase might be seen as a disadvantage, but adding new tests is simply a natural consequence of introducing new source code. Even if the logic were added to a microservice, tests would still be required there. In every case, tests are not a disadvantage but an inevitable part of the decision. I also hope we have tests in place for that.

Reviewer 2: it might be the wording is not specific, but I think from what I gather here, the disadvantage is "more complexity in testing" not testing or writing tests in general. or am I wrong? 🙏 🤲

Author: I guess you are right Pram :) 

would you be fine with this: 
    - More complexity in testing
      - Tests to test the new methods and schemas
      - Tests to check that the result is a valid XJustiz message (may require a mini microservice)

Here I left the tests out that also have to be done using the library (the tests to check if our data ist reflected in the standard)

Reviewer 2: "More complexity in testing" is fine

Author: alright I updated it accordingly

      - Tests to check that our data is reflected in the standards
      - Tests to test the new methods and schemas

>       - tests to check that the result is a valid XJusitz message (may require a mini microservice)

Reviewer 2: ```suggestion
      - tests to check that the result is a valid XJustiz message (may require a mini microservice)
```

>     - We will have to maintain at least two different XJusitz versions at the same time for interoperability.

Reviewer 2: ```suggestion
    - We will have to maintain at least two different XJustiz versions at the same time for interoperability.
```

>   - If the XJusitz version changes:

Reviewer 2: ```suggestion
  - If the XJustiz version changes:
```

That would need a new service with according interface in our cluster which entails deployment and maintenance effort.
Also, we create two very distinct places to handle business logic.

- Advantages:

  - A well-maintained library
  - could be used as a microservice:

>     - XJustiz versioning would be easier because it can be done with less manual customisation.

Author: My idea behind this was that more things will happen automatically with the jaxb library (the only thing that would also have to be updated "manually" are the codelists [but probably we will find some automation for the codelists for both the TS and jaxb case]), but all the classes etc. will be generated automatically with the new XSD schemas, without having to find the changes manually.

>     - XJustiz versioning would be easier because it can be done with less manual customisation.

Reviewer: Aahh okay, so this is about the new version? Can we rephrase like this maybe?
```suggestion
    - adapting to new XJustiz versions would be easier because it can be done with less manual customisation.
```
And then probably move this out of the "microservice" super-point, as I think this is more a lib advantage than microservice?

>     - XJustiz versioning would be easier because it can be done with less manual customisation.

Author: good idea :)

    - Adds less complexity to the current code base
    - KomPla could also use the service in the future.
    - No need for new functions/methods to implement types like selections as they are part of the library

>       - Tests for those new methods/functions are not needed as they are part of the library.

Author: I think I also did not formulate it that well - what I mean is the following:
example Nachrichtenkopf needs Empfänger which has 4 options of which exactly one has to be chosen (choice element in xml) 

In the generated TS case it looks like this:
```
interface _TypeGDSNachrichtenkopf extends BaseType {
xjustizVersion: string;
aktenzeichenabsender?: string[];
aktenzeichenempfaenger?: string[];
auswahl_absender: TypeGDSNachrichtenkopfAuswahl_absenderType;
auswahl_empfaenger: TypeGDSNachrichtenkopfAuswahl_empfaengerType;
}

interface _TypeGDSNachrichtenkopfAuswahl_empfaengerType extends BaseType {
	empfaengergericht: CodeGDSGerichteTyp3;
	empfaengerrvTraeger: CodeGDSRVTraeger;
	empfaengersonstige: string;
}
```

in the generated java code like this (the fields in java and ts are in different orders - I just copied a part of them until I got to the important part which is the empaenger): 

```
public class TypeGDSNachrichtenkopf {
    @XmlElement(name = "aktenzeichen.absender")
    protected List<String> aktenzeichenAbsender;
    @XmlElement(name = "aktenzeichen.empfaenger")
    protected List<String> aktenzeichenEmpfaenger;
    @XmlElement(required = true)
    @XmlSchemaType(name = "dateTime")
    protected XMLGregorianCalendar erstellungszeitpunkt;
    @XmlElement(name = "auswahl_absender", required = true)
    protected TypeGDSNachrichtenkopf.AuswahlAbsender auswahlAbsender;
    @XmlElement(name = "auswahl_empfaenger", required = true)
    protected TypeGDSNachrichtenkopf.AuswahlEmpfaenger auswahlEmpfaenger;`

@XmlAccessorType(XmlAccessType.FIELD)
    @XmlType(name = "", propOrder = {
        "absenderGericht",
        "absenderSonstige",
        "absenderRvTraeger",
        "absenderPolizeibehoerde"
    })
    public static class AuswahlAbsender {

        @XmlElement(name = "absender.gericht")
        protected CodeGDSGerichteTyp3 absenderGericht;
        @XmlElement(name = "absender.sonstige")
        protected String absenderSonstige;
        @XmlElement(name = "absender.rvTraeger")
        protected CodeGDSRVTraeger absenderRvTraeger;
        @XmlElement(name = "absender.polizeibehoerde")
        protected CodeGDSPolizeibehoerdenTyp3 absenderPolizeibehoerde;

```
  
        
     So if I am not mistaken there are two problems in TS that don't exist in Java: 
     1. in TS all Empänger fields are required while only one has to be chosen 
     2. If I set them all to optional then I can just add an empty empfänger to Nachrichtenkopf without getting an error.
     
     In jaxb you can create an empfeanger without setting fields but it will then automatically throw an error when using the inbuilt validation (here the error for a simple Basisnachricht where all the other required fields are there: "cvc-complex-type.2.4.a: Invalid content was found starting with element '{"http://www.xjustiz.de":auswahl_empfaenger}'. One of '{"http://www.xjustiz.de":auswahl_absender}' is expected.")

>     - Tests to check if a message is a valid XJustiz message are not needed as this is part of the library

Author: What I mean is that it can automatically check if all required fields of an XJusitz Nachricht exist (see comment above)

    - Clear separation of services (user flow, xml generation, sending data with FITKO)

>     - With FITKO Java is already part of our project

Author: good idea :)

>     - Once we send the xml, changes to the user data will probably occur very rarely (we can only send data that is reflected in the xsd, and the xsd is [in theory] a representation of the required data).

Author: I think I didn't express my thoughts very well because I somehow put two thoughts into one: 

1. The first idea behind this was that in our case (the flight right case), for example, we are currently still making a lot of changes to the data because of the ongoing content review. However, when the content review is done and when we are launching there will probably not be (or very rarely) any changes to the data. Thus, if we would only define the final API at the very end after the content reviews etc. are done then the interoperability issues might not occur so often

2. The second thought is actually less important but somehow got more space in the description :D 
The changes we make to our data should always be able to be reflected in the schema. The idea is that we should not change data or add data that we cannot end up sending because it is not part of the standard. This would also mean more alignment to the standard from the beginning and less interoperability issues in the end 
- but admittedly less important then the first point and I see how it made the whole thing a bit confusing ;)

>     - Once we send the xml, changes to the user data will probably occur very rarely (we can only send data that is reflected in the xsd, and the xsd is [in theory] a representation of the required data).

Reviewer: aaahhh nice!
Suggestion for formulating this:
```suggestion
    - Interoperability problems may be less bad, as changes to the user data schema will probably occur very rarely (after having the first version live). Also, when changes happen to the userData schema, they in a lot of cases are triggered from changes to the xsd schema, reuslting in a new version anyways.
```

Maaayyybe? Second part probably needs some work still^^

>     - Once we send the xml, changes to the user data will probably occur very rarely (we can only send data that is reflected in the xsd, and the xsd is [in theory] a representation of the required data).

Author: Nice - makes more sense :) 
reagrding the second part maybe:
Furthermore, since the XJusitz standard should serve as a representation of the necessary data for a specific use case, the possible changes to the userData schema are limited and might rather be the result of changes in the XJusitz standard, which would result in a new version anyway.

>     - Once we send the xml, changes to the user data will probably occur very rarely (we can only send data that is reflected in the xsd, and the xsd is [in theory] a representation of the required data).

Reviewer: Nice 👌

- Disadvantages:
  - Microservice can be an overhead:

>     - Requires additional infrastructure

Author: I will add it :)

That would need a new service with according interface in our cluster which entails deployment and maintenance effort.
Also, we create two very distinct places to handle business logic.

- Advantages:

  - A well-maintained library
  - could be used as a microservice:

>     - XJusitz versioning would be easier because it can be done with less manual customisation.

Reviewer 2: ```suggestion
    - XJustiz versioning would be easier because it can be done with less manual customisation.
```

    - Adds less complexity to the current code base
    - KomPla could also use the service in the future.
    - No need for new functions/methods to implement types like selections as they are part of the library
      - Tests for those new methods/functions are not needed as they are part of the library.

>     - Tests to check if a message is a valid XJusitz message are not needed as this is part of the library

Reviewer 2: ```suggestion
    - Tests to check if a message is a valid XJustiz message are not needed as this is part of the library
```
